command/agent/gated_writer.go

```diff
@@ -29,8 +29,8 @@ func (w *GatedWriter) Flush() {
 }
 
 func (w *GatedWriter) Write(p []byte) (n int, err error) {
-	w.lock.RLock()
```

Comment:
Hi @WIZARD-CXY thanks for the PR!

I think the read lock was intended to not synchronize writers after the flush was done (since there's no way to set `flush` to `false` again, and after startup it will basically be going through that path forever). You are right there's a race here, though. I think this fix would have the intended behavior:

```go
func (w *GatedWriter) Write(p []byte) (n int, err error) {
        // Once we flush we no longer synchronize writers since there's
        // no use of the internal buffer. This is the happy path.
        w.lock.RLock()
        if w.flush {
                w.lock.RUnlock()
                return w.Writer.Write(p)
        }
        w.lock.RUnlock()

        // Now take the write lock.
        w.lock.Lock()
        defer w.lock.Unlock()

        // Things could have changed between the locking operations, so we
        // have to check one more time.
        if w.flush {
                return w.Writer.Write(p)
        }

        // Buffer up the written data.
        p2 := make([]byte, len(p))
        copy(p2, p)
        w.buf = append(w.buf, p2)
        return len(p), nil
}
```